Added unit tests and a Gruntfile

wolfeman67 opened the pull request:
The Gruntfile runs jshint and the test file that is located in the tests folder.
The tests include checking:
1. If a bridge is created when there are none to be found.
2. If a bridge is used again if a preexisting one is found.
3. If a channel is allowed to continue (from originating) when a valid endpoint is given.
4. If a channel isn't allowed to continue (from originating) when an invalid endpoint is given.

wolfeman67 committed:
Added unit tests and a Gruntfile

The Gruntfile runs jshint and the test file that is located in the tests folder.
The tests include checking:
1. If a bridge is created when there are none to be found.
2. If a bridge is used again if a preexisting one is found.
3. If a channel is allowed to continue (from originating) when a valid endpoint is given.
4. If a channel isn't allowed to continue (from originating) when an invalid endpoint is given.

wolfeman67 committed:
.Gitignore file

samuelg commented on Gruntfile.js:
```diff
@@ -0,0 +1,20 @@
+module.exports = function(grunt) {
+   grunt.initConfig({
+       jshint: {
+          files: ['Gruntfile.js', 'test/**/*.js', 'lib/*.js', '*.js']
```
If you use the /*_/_.js syntax you should also apply it to lib/. Instead of specifying which subdirectories you care about, you are saying to check all subdirectories under test/ or lib/.

samuelg commented on Gruntfile.js:
```diff
@@ -0,0 +1,20 @@
+    grunt.loadNpmTasks('grunt-mocha-test');
+
+    grunt.registerTask('default', ['jshint', 'mochaTest']);
+    };
```
Several lines here are not aligned correctly and makes the file hard to read. This close }; should match the indentation of the module.exports line. The grunt.loadNpmTasks lines should match the indentation of the grunt.initConfig line. The closing }}); before the first grunt.loadNpmTests line should match the indentation of the grunt.initConfig line as well.

samuelg commented on Gruntfile.js:
```diff
@@ -0,0 +1,20 @@
+module.exports = function(grunt) {
+   grunt.initConfig({
+       jshint: {
```
You have jshint configured but you have no .jshintrc file to drive the behavior, see https://raw.githubusercontent.com/asterisk/node-voicemail-data/master/.jshintrc for an example of how to configure jshint.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+var pastStasis = 0;
+
+// The mocked up version of the callback error function
+var end= function(err){console.log(err);}
```
should be a space after end and after the closing ) of the function signature. I would also move console.log on the next line.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+    bridge_channels = [];
+    bridge_created = 0;
+    done();
+    });
```
This line should match the indentation of the afterEach line.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+  it('should create a bridge when there isn\'t one', function(done) {
+    var channel = getMockChannel();
+    var client = getMockClient();
+    var sla = require('../lib/sla.js')(client, end);
```
Calling the sla module this way with an error callback that simply logs the error means you can't test the module for failure. You should be able to test for a failure in the sla module using different mocks or mocks configured diffently.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+describe('SLA Bridge and Channel Tester', function() {
+
+  afterEach(function(done) {
+    channels = [];
```
You should also reset bridges, valid, and pastStasis.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+var Emitter = require('events').EventEmitter;
+
+var mockClient;
+var mockChannel;
```
I don't think you need this since you are storing mock channels for a given test in the channels array.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+
+var mockClient;
+var mockChannel;
+var mockBridge;
```
I don't think you need this since you are storing mock bridges for a given test in the bridges array.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+        return bridges;
+      },
+      create: function(param, cb) {
+        bridge_created = 1;
```
Couldn't you just check the bridges array to see if a new bridge was created?

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+        console.log(param);
+        bridges.push(getMockBridge(param));
+        console.log(bridges)
+        cb(null, bridges[0]);
```
This will always pass the first bridge that was created instead of the latest. Use bridges[bridge.length - 1] instead.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+    };
+    this.Channel= function() {
+      return getMockChannel();
+      }
```
This should be aligned with the this.Channel line.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+      return getMockChannel();
+      }
+  };
+  mockClient = new Client();
```
You should be making this an EventEmitter so that you can emit events to the sla module.

wolfeman67 replied:
Will do later when this becomes necessary. For now, it is getting skipped.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+  var Bridge = function(param) {
+    console.log(param.type);
+    this.bridge_type = param.type;
+    this.id = "1";
```
You should increment ids so more than one bridge can be created with unique IDs.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+    this.bridge_type = param.type;
+    this.id = "1";
+    this.addChannel= function(input, cb) {
+      console.log("TREX");
```
console.log statements are ok while you are writing tests but generally do not get included as part of the tests. Tests report whether they pass or not to the terminal but usually no more.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+    this.id = "1";
+    this.addChannel= function(input, cb) {
+      console.log("TREX");
+      pastStasis = 1;
```
I'm not clear what this is for?

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+      console.log("TREX");
+      pastStasis = 1;
+      var channel = channels.filter(function(testChan) {
+        return testChan.id === '1';
```
the channel will be passed to the method addChannel, you shoudn't get a channel this way to "add" it to the bridge. In general, your mock should work like the real thing. If it expects a channel in the arguments, you should use that.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+      bridge_channels.push(channel);
+      cb(null);
+      console.log(channel);
+      this.emit('ChannelEnteredBridge', {channel: {id: channel.id}});
```
the event object passed to the event should also include info on the bridge. Additionally, a second object should be passed that include a channel and a bridge key with mock objects as values.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+      setTimeout(function() {
+        if(valid == 1) {
+          console.log(self);
+          self.emit('StasisStart', self, self);
```
The first object passed to the event should not simply be a channel but rather an event object that includes a channel key with data about the channel. If a library you are testing looks at event.channel.id for example it would fail.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+
+var getMockChannel = function() {
+  var Channel = function() {
+    this.id = '1';
```
Use incrementing ids so you can create more than 1 mock channel with unique IDs.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+          console.log(self);
+          self.emit('StasisStart', self, self);
+        }
+      }, 20);
```
You should use an async delay variable instead of hardcoding this.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+    done();
+    });
+
+  it('should create a bridge when there isn\'t one', function(done) {
```
Since sla.js is supposed to create a mixing bridge, you should validate that this is what occurs.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+    } 
+  });
+
+  it('should use a preexisting bridge if there is one', function(done) {
```
This test appears to depend on the previous one running first, this may lead to problems in the future. Tests should be independent of each other. Create a bridge first and then import the sla module to test this. afterEach should reset all state related vars in this file.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+    bridgeChecking();
+    function bridgeChecking() {
+      setTimeout(function() {
+        if(channels.length != 0 && bridge_created == 0) {
```
This could pass even though the bridge could be about to be created. You may want to keep track of the following function in sla.js being called before you check if a bridge was created.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+    } 
+  });
+
+  it('should use a "valid" endpoint and give off a StasisStart event ' +
```
Since this first task involves hard coding endpoints, you may want to specifically check for the expected input.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+    } 
+  });
+
+  it('should use an "invalid" endpoint and not give off a StasisStart event',
```
This test is no different from the previous one, it's only passing because the check occurs before the call updates the variables you are checking. Until sla.js is made to expect parameters to drive what endpoint is called you may not be able to test this.

samuelg commented:
You should have grunt and all grunt plugins in your dev-dependencies in package.json. Currently grunt does not run after a checkout and npm install.

samuelg commented:
running jshint on tests/sample.js found the following errors:

tests/sample.js: line 5, col 1, Use the function form of "use strict".
tests/sample.js: line 30, col 42, Missing semicolon.
tests/sample.js: line 55, col 29, Missing semicolon.
tests/sample.js: line 61, col 8, Missing semicolon.
tests/sample.js: line 136, col 28, Use '!==' to compare with '0'.
tests/sample.js: line 153, col 28, Use '!==' to compare with '0'.
tests/sample.js: line 153, col 51, Use '===' to compare with '0'.
tests/sample.js: line 171, col 54, Use '!==' to compare with '0'.
tests/sample.js: line 191, col 23, Use '===' to compare with '0'.
tests/sample.js: line 191, col 54, Use '===' to compare with '0'.
tests/sample.js: line 10, col 12, 'require' is not defined.
tests/sample.js: line 11, col 15, 'require' is not defined.
tests/sample.js: line 30, col 24, 'console' is not defined.
tests/sample.js: line 47, col 9, 'console' is not defined.
tests/sample.js: line 53, col 9, 'console' is not defined.
tests/sample.js: line 55, col 9, 'console' is not defined.
tests/sample.js: line 74, col 5, 'console' is not defined.
tests/sample.js: line 78, col 7, 'console' is not defined.
tests/sample.js: line 85, col 7, 'console' is not defined.
tests/sample.js: line 103, col 7, 'console' is not defined.
tests/sample.js: line 105, col 7, 'setTimeout' is not defined.
tests/sample.js: line 107, col 11, 'console' is not defined.
tests/sample.js: line 128, col 3, 'it' is not defined.
tests/sample.js: line 131, col 15, 'require' is not defined.
tests/sample.js: line 135, col 7, 'setTimeout' is not defined.
tests/sample.js: line 145, col 3, 'it' is not defined.
tests/sample.js: line 148, col 15, 'require' is not defined.
tests/sample.js: line 152, col 7, 'setTimeout' is not defined.
tests/sample.js: line 162, col 3, 'it' is not defined.
tests/sample.js: line 166, col 15, 'require' is not defined.
tests/sample.js: line 170, col 7, 'setTimeout' is not defined.
tests/sample.js: line 181, col 3, 'it' is not defined.
tests/sample.js: line 186, col 15, 'require' is not defined.
tests/sample.js: line 190, col 7, 'setTimeout' is not defined.

Some of these are due to your missing a .jshintrc file.

samuelg commented on tests/sample.js:
```diff
@@ -0,0 +1,199 @@
+/**
```
Why is this file name sample.js? If this will be the only test file (small module testing), it should be called tests.js. If you will be testing multiple modules without the repo, it should probably be called sla.js.

wolfeman67 committed:
Fixed miscellanenous issues related to Gruntfile and test file

Fixed jshint issues, dev-dependencies, cleaned up the test, and made sure
that the tests now account for the code to be using promises.

wolfeman67 committed:
Fixed mostly jshint setup errors and made conditionals actually boolean values

samuelg commented on tests/sla_tests.js:
```diff
@@ -0,0 +1,212 @@
+    bridgeChecking();
+    function bridgeChecking() {
+      setTimeout(function() {
+        if(channels.length !== 0 && usingExisting === false &&
```
since if statements already test for truth, you can just use  !usingExisting instead of usingExisting === false and isMixing instead of isMixing === true. Also, because 0 evaluates to false, you can just use channels.length instead of channels.length !== 0 but you can leave that if you want to make it explicit.

samuelg commented on tests/sla_tests.js:
```diff
@@ -0,0 +1,212 @@
+    this.addChannel= function(input, cb) {
+      //Here channels and bridges should have the same ID number
+      var channel = channels.filter(function(testChan) {
+        return testChan.id === input.id;
```
You have not addressed this yet. The input to addChannel looks like {channel: 'id'} and not {id: 'id'}.

wolfeman67 committed:
Made the boolean statements more "correct" and fixed input.channel mixup

samuelg commented:
Looks good. Since you have a lot of commits to address PR findings, I would suggest we walk though squashing the commits into a single commit in a different branch before merging.

wolfeman67 commented:
Squashed into #7
